In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# Raw data file
file_to_load = "Resources/purchase_data.csv"

# Read purchasing file and store into pandas data frame
purchase_data_df = pd.read_csv(file_to_load)

#print intial summary head() table
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
#Check for rows of missing data in each column of the data frame
purchase_data_df.count()
#Returns no rows missing data

Purchase ID    780
SN             780
Age            780
Gender         780
Item ID        780
Item Name      780
Price          780
dtype: int64

In [3]:
#list names of columns as they appear in the background for possible differences
list(purchase_data_df)
#names of columns appear as they do in the intial summary table above

['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price']

In [4]:
#Count the number of times a unique player ID made a purchase and therefore have the total number of unique players
unique_player_count = purchase_data_df['SN'].nunique()

#display number of unique players
unique_player_count

576

In [5]:
#Convert the unique_player_count into a dictionary
unique_player_dicts = [{"Total Players": (unique_player_count)}]

#Convert unique_player_dicts dictionary into a data frame
total_players_df = pd.DataFrame(unique_player_dicts)

#Display the number_of_players_df
total_players_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [6]:
#print intial summary head() table for reference
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [7]:
#Display a statistical overview of the data frame
purchase_data_df.describe()
#Can pull number of purchases and average price data from Purchase ID count and Price mean:

Purchase ID         Age     Item ID       Price
count   780.000000  780.000000  780.000000  780.000000
mean    389.500000   22.714103   92.114103    3.050987
std     225.310896    6.659444   52.775943    1.169549
min       0.000000    7.000000    0.000000    1.000000
25%     194.750000   20.000000   48.000000    1.980000
50%     389.500000   22.000000   93.000000    3.150000
75%     584.250000   25.000000  139.000000    4.080000
max     779.000000   45.000000  183.000000    4.990000

In [8]:
#Count the number unique item IDs and therefore have the total number of unique items purchased
unique_items_purchased = purchase_data_df['Item ID'].nunique()

#display number of unique players and pull that information for summary data frame
unique_items_purchased

183

In [9]:
#Sum the total price of each purchase in column price and therefore have the total revenue
total_revenue = purchase_data_df['Price'].sum()

#display the total revenue
total_revenue

2379.77

In [10]:
#Create a dictionary for the information desired
purchasing_analysis_dicts = [{"Number of Unique Items":(unique_items_purchased), "Average Price": "3.05",
                              "Number of Purchases": "780", "Total Revenue": (total_revenue)}]

#Convert dictionary into a data frame
purchasing_analysis_df = pd.DataFrame(purchasing_analysis_dicts)

# Reorganize the columns
purchasing_analysis_df = purchasing_analysis_df[["Number of Unique Items","Average Price","Number of Purchases","Total Revenue"]]

purchasing_analysis_df["Average Price"]=purchasing_analysis_df["Average Price"].astype(float)

purchasing_analysis_df["Average Price"]=purchasing_analysis_df["Average Price"].map('${:,.2f}'.format)

purchasing_analysis_df["Total Revenue"]=purchasing_analysis_df["Total Revenue"].map('${:,.2f}'.format)

#Display the purchasing_analysis_df
purchasing_analysis_df

Number of Unique Items Average Price Number of Purchases Total Revenue
0                     183         $3.05                 780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [11]:
#print intial summary head() table for reference
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [12]:
# Calculations can also be performed on Series and added into DataFrames as new columns
gender_breakdown = purchase_data_df["Gender"].value_counts()

gender_breakdown_df = pd.DataFrame(gender_breakdown)

gender_breakdown_df.head()

Gender
Male                      652
Female                    113
Other / Non-Disclosed      15

In [13]:
# Convert the column name into "Total Count"
gender_breakdown_df = gender_breakdown_df.rename(
    columns={"Gender": "Total Count"})

#Display new data frame
gender_breakdown_df.head()

Total Count
Male                           652
Female                         113
Other / Non-Disclosed           15

In [14]:
#Calculate the percentage of players by gender
percentage_of_players = gender_breakdown_df["Total Count"]/total_players_df["Total Players"].sum() * 100

#Create a new column to hold percentage_of_players
gender_breakdown_df["Percentage of Players"] = percentage_of_players

#Display new data frame
gender_breakdown_df.head()

Total Count  Percentage of Players
Male                           652             113.194444
Female                         113              19.618056
Other / Non-Disclosed           15               2.604167

In [15]:
# Reorganize the columns
gender_breakdown_df = gender_breakdown_df[["Percentage of Players","Total Count"]]

# Change the formatting of the "percentage of players" columns to two decimals and add a %
gender_breakdown_df["Percentage of Players"] = gender_breakdown_df["Percentage of Players"].map('{:,.2f}%'.format)

#Display new data frame
gender_breakdown_df.head()

Percentage of Players  Total Count
Male                                113.19%          652
Female                               19.62%          113
Other / Non-Disclosed                 2.60%           15


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [16]:
#create a groupby data frame for males

purchase_data_v1_df = purchase_data_df.loc[purchase_data_df["Gender"] == "Male", :]

purchase_data_v1_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [17]:
#create a groupby data frame for females

purchase_data_v2_df = purchase_data_df.loc[purchase_data_df["Gender"] == "Female", :]

purchase_data_v2_df.head()

Purchase ID         SN  Age  Gender  Item ID  \
15           15  Lisassa64   21  Female       98   
18           18  Reunasu60   22  Female       82   
38           38   Reulae52   10  Female      116   
41           41  Assosia88   20  Female        7   
55           55  Phaelap26   25  Female       84   

                       Item Name  Price  
15   Deadline, Voice Of Subtlety   2.89  
18                       Nirvana   4.90  
38       Renewed Skeletal Katana   4.18  
41  Thorn, Satchel of Dark Souls   1.33  
55                    Arcane Gem   3.79

In [18]:
#create a groupby data frame for Other / Non-Disclosed

purchase_data_v3_df = purchase_data_df.loc[purchase_data_df["Gender"] == "Other / Non-Disclosed", :]

purchase_data_v3_df.head()

Purchase ID            SN  Age                 Gender  Item ID  \
9              9   Chanosian48   35  Other / Non-Disclosed      136   
22            22  Siarithria38   38  Other / Non-Disclosed       24   
82            82    Haerithp41   16  Other / Non-Disclosed      160   
111          111      Sundim98   21  Other / Non-Disclosed       41   
228          228   Jiskirran77   20  Other / Non-Disclosed       80   

                        Item Name  Price  
9    Ghastly Adamantite Protector   3.58  
22                  Warped Fetish   3.81  
82                     Azurewrath   4.40  
111                         Orbit   4.75  
228                     Dreamsong   3.39

In [19]:
#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender

male_gender_count = purchase_data_v1_df["Gender"].count()

male_total_purchase_value = purchase_data_v1_df["Price"].sum()

male_average_purchase_price = purchase_data_v1_df["Price"].mean()

male_average_purchase_price_per_person = purchase_data_v1_df["Price"].mean()

female_gender_count = purchase_data_v2_df["Gender"].count()

female_total_purchase_value = purchase_data_v2_df["Price"].sum()

female_average_purchase_price = purchase_data_v2_df["Price"].mean()

female_average_purchase_price_per_person = purchase_data_v2_df["Price"].mean()

other_gender_count = purchase_data_v3_df["Gender"].count()

other_total_purchase_value = purchase_data_v3_df["Price"].sum()

other_average_purchase_price = purchase_data_v3_df["Price"].mean()

other_average_purchase_price_per_person = purchase_data_v3_df["Price"].mean()



#df = pd.DataFrame(
              #{"a" : [4 ,5, 6],
               #"b" : [7, 8, 9],
               #"c" : [10, 11, 12]},
            #index = [1, 2, 3])

In [20]:
#Generate a data frame using the above variables using the following format
#df = pd.DataFrame(
              #{"a" : [4 ,5, 6],
               #"b" : [7, 8, 9],
               #"c" : [10, 11, 12]},
            #index = [1, 2, 3])
purchasing_analysis_by_gender_df = pd.DataFrame(
                {"Purchase Count" : ["", (female_gender_count), (male_gender_count), (other_gender_count)],
               "Average Purchase Price" : ["", (female_average_purchase_price), (male_average_purchase_price), (other_average_purchase_price)],
               "Total Purchase Value" : ["", (female_total_purchase_value), (male_total_purchase_value), (other_total_purchase_value)],
               "Avg Purchase Total Per Person" : ["", (female_average_purchase_price_per_person), (male_average_purchase_price_per_person), (purchase_data_v3_df["Price"].mean())] 
                },
            index = ["Gender", "Female", "Male", "Other / Non-Disclosed"])
purchasing_analysis_by_gender_df.head()

Purchase Count Average Purchase Price  \
Gender                                                        
Female                           113                3.20301   
Male                             652                3.01785   
Other / Non-Disclosed             15                  3.346   

                      Total Purchase Value Avg Purchase Total Per Person  
Gender                                                                    
Female                              361.94                       3.20301  
Male                               1967.64                       3.01785  
Other / Non-Disclosed                50.19                         3.346

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [21]:
# Establish bins for ages
age_bins = [0, 9.90, 14.90, 19.90, 24.90, 29.90, 34.90, 39.90, 99999]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

In [22]:
#Categorize the existing players using the age bins
purchase_data_df["Age Range"] = pd.cut(purchase_data_df["Age"], age_bins, labels=group_names)
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Range  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [23]:
#Drop duplicate rows by locating the duplicate gamer IDs and dropping them
purchase_data_df_dropped = purchase_data_df.drop_duplicates(["SN"])
purchase_data_df_dropped.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Range  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [24]:
#Calculate the numbers and percentages by age group
age_range_count = purchase_data_df_dropped["Age Range"].value_counts()
age_range_count

20-24    258
15-19    107
25-29     77
30-34     52
35-39     31
10-14     22
<10       17
40+       12
Name: Age Range, dtype: int64

In [25]:
#Calculate the numbers and percentages by age group
pct_players = age_range_count/purchase_data_df_dropped["SN"].count() * 100

pct_players

20-24    44.791667
15-19    18.576389
25-29    13.368056
30-34     9.027778
35-39     5.381944
10-14     3.819444
<10       2.951389
40+       2.083333
Name: Age Range, dtype: float64

In [26]:
#Create a summary data frame to hold the results
age_demo_df = pd.DataFrame({"Percentage of Players" : pct_players, "Total Count" : age_range_count})

age_demo_df.sort_index().round(2)

Percentage of Players  Total Count
<10                     2.95           17
10-14                   3.82           22
15-19                  18.58          107
20-24                  44.79          258
25-29                  13.37           77
30-34                   9.03           52
35-39                   5.38           31
40+                     2.08           12

In [27]:
# Change the formatting of the "percentage of players" columns to two decimals and add a %
age_demo_df["Percentage of Players"] = age_demo_df["Percentage of Players"].map('{:,.2f}%'.format)

In [28]:
#Display Final Formatted DataFrame
age_demo_df.sort_index()

Percentage of Players  Total Count
<10                   2.95%           17
10-14                 3.82%           22
15-19                18.58%          107
20-24                44.79%          258
25-29                13.37%           77
30-34                 9.03%           52
35-39                 5.38%           31
40+                   2.08%           12

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [29]:
#Bin the purchase_data data frame by age
purchase_data_df["Age Range"] = pd.cut(purchase_data_df["Age"], age_bins, labels=group_names)
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Range  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [30]:
#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
total_purchase_value = purchase_data_df.groupby(["Age Range"]).sum()["Price"]
total_purchase_value

Age Range
<10        77.13
10-14      82.78
15-19     412.89
20-24    1114.06
25-29     293.00
30-34     214.00
35-39     147.67
40+        38.24
Name: Price, dtype: float64

In [31]:
#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
avg_purchase_price = purchase_data_df.groupby(["Age Range"]).mean()["Price"]
avg_purchase_price

Age Range
<10      3.353478
10-14    2.956429
15-19    3.035956
20-24    3.052219
25-29    2.900990
30-34    2.931507
35-39    3.601707
40+      2.941538
Name: Price, dtype: float64

In [32]:
#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
purchase_count = purchase_data_df.groupby(["Age Range"]).count()["Price"]
purchase_count

Age Range
<10       23
10-14     28
15-19    136
20-24    365
25-29    101
30-34     73
35-39     41
40+       13
Name: Price, dtype: int64

In [33]:
#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
avg_purchase_per_person = total_purchase_value/purchase_count
avg_purchase_per_person

Age Range
<10      3.353478
10-14    2.956429
15-19    3.035956
20-24    3.052219
25-29    2.900990
30-34    2.931507
35-39    3.601707
40+      2.941538
Name: Price, dtype: float64

In [34]:
#Create a data frame based on the above calculated variables

age_purchases_df = pd.DataFrame({"Purchase Count" : purchase_count, "Average Purchase Price" : avg_purchase_price,
                                "Total Purchase Value": total_purchase_value, "Average Purchase Total Per Person": avg_purchase_per_person})
age_purchases_df.sort_index().round(2)

Purchase Count  Average Purchase Price  Total Purchase Value  \
Age Range                                                                 
<10                    23                    3.35                 77.13   
10-14                  28                    2.96                 82.78   
15-19                 136                    3.04                412.89   
20-24                 365                    3.05               1114.06   
25-29                 101                    2.90                293.00   
30-34                  73                    2.93                214.00   
35-39                  41                    3.60                147.67   
40+                    13                    2.94                 38.24   

           Average Purchase Total Per Person  
Age Range                                     
<10                                     3.35  
10-14                                   2.96  
15-19                                   3.04  
20-24                                   3.05  
25-29                                   2.90  
30-34                                   2.93  
35-39                                   3.60  
40+                                     2.94

In [35]:
# Change the formatting - add a $ to the columns that need it.
age_purchases_df["Average Purchase Price"]=age_purchases_df["Average Purchase Price"].map('${:,.2f}'.format)
age_purchases_df["Total Purchase Value"]=age_purchases_df["Total Purchase Value"].map('${:,.2f}'.format)
age_purchases_df["Average Purchase Total Per Person"]=age_purchases_df["Average Purchase Total Per Person"].map('${:,.2f}'.format)

In [36]:
#Display Final Formatted DataFrame
age_purchases_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Range                                                               
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

          Average Purchase Total Per Person  
Age Range                                    
<10                                   $3.35  
10-14                                 $2.96  
15-19                                 $3.04  
20-24                                 $3.05  
25-29                                 $2.90  
30-34                                 $2.93  
35-39                                 $3.60  
40+                                   $2.94

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [37]:
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Range  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [38]:
sn_groupby_df = purchase_data_df.groupby(['SN'])
sn_groupby_df

In [39]:
sn_purchase_count = sn_groupby_df["Purchase ID"].count()

sn_purchase_count.head()

SN
Adairialis76    1
Adastirin33     1
Aeda94          1
Aela59          1
Aelaria33       1
Name: Purchase ID, dtype: int64

In [40]:
sn_purchase_avg = sn_groupby_df["Price"].mean()

sn_purchase_avg.head()

SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [41]:
sn_purchase_total = sn_groupby_df["Price"].sum()

sn_purchase_total.head()

SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [42]:
top_spenders_df = pd.DataFrame({"Purchase Count" : sn_purchase_count, "Average Purchase Price" : sn_purchase_avg,
                                "Total Purchase Value": sn_purchase_total})
top_spenders_df.sort_index().round(2)

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                           
Adairialis76                  1                    2.28                  2.28
Adastirin33                   1                    4.48                  4.48
Aeda94                        1                    4.91                  4.91
Aela59                        1                    4.32                  4.32
Aelaria33                     1                    1.79                  1.79
Aelastirin39                  2                    3.64                  7.29
Aelidru27                     1                    1.09                  1.09
Aelin32                       3                    2.99                  8.98
Aelly27                       2                    3.39                  6.79
Aellynun67                    1                    3.74                  3.74
Aellyria80                    1                    3.08                  3.08
Aelollo59                     2                    2.82                  5.63
Aenarap34                     1                    2.96                  2.96
Aeral43                       1                    4.40                  4.40
Aeral68                       1                    4.00                  4.00
Aeral97                       1                    4.80                  4.80
Aeralria27                    1                    4.09                  4.09
Aeralstical35                 1                    2.96                  2.96
Aeri84                        1                    1.61                  1.61
Aerillorin70                  1                    3.33                  3.33
Aerithllora36                 2                    4.32                  8.64
Aerithnucal56                 1                    4.40                  4.40
Aerithnuphos61                1                    4.91                  4.91
Aerithriaphos45               1                    1.56                  1.56
Aerithriaphos46               1                    2.18                  2.18
Aesri53                       1                    1.76                  1.76
Aesty53                       2                    1.96                  3.91
Aestysu37                     2                    2.69                  5.38
Aesur96                       1                    2.21                  2.21
Aesurstilis64                 1                    1.03                  1.03
...                         ...                     ...                   ...
Undosia27                     1                    3.77                  3.77
Undosian34                    2                    4.15                  8.30
Undotesta33                   1                    4.74                  4.74
Wailin72                      1                    2.73                  2.73
Yadacal26                     2                    3.27                  6.54
Yadaisuir65                   1                    4.09                  4.09
Yadam35                       1                    2.48                  2.48
Yadanu52                      1                    2.38                  2.38
Yadaphos40                    2                    2.68                  5.35
Yalae81                       2                    3.34                  6.69
Yalaeria91                    1                    1.06                  1.06
Yaliru88                      1                    3.03                  3.03
Yalo85                        1                    3.65                  3.65
Yalostiphos68                 1                    4.40                  4.40
Yana46                        1                    3.55                  3.55
Yarithllodeu72                1                    3.75                  3.75
Yarithrgue83                  1                    3.45                  3.45
Yarithrin84                   1                    4.14                  4.14
Yarithsurgue62                1        

In [43]:
#Sort the total purchase value column in descending order
top_spenders_df = top_spenders_df.sort_values(by='Total Purchase Value', ascending=False)
top_spenders_df.head()

Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                       
Lisosia93                 5                3.792000                 18.96
Idastidru52               4                3.862500                 15.45
Chamjask73                3                4.610000                 13.83
Iral74                    4                3.405000                 13.62
Iskadarya95               3                4.366667                 13.10

In [44]:
# Change the formatting of the columns to two decimals and add a $
top_spenders_df["Average Purchase Price"]=top_spenders_df["Average Purchase Price"].map('${:,.2f}'.format)
top_spenders_df["Total Purchase Value"]=top_spenders_df["Total Purchase Value"].map('${:,.2f}'.format)

In [45]:
#Display Final Formatted DataFrame
top_spenders_df

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                       
Lisosia93                   5                  $3.79               $18.96
Idastidru52                 4                  $3.86               $15.45
Chamjask73                  3                  $4.61               $13.83
Iral74                      4                  $3.40               $13.62
Iskadarya95                 3                  $4.37               $13.10
Ilarin91                    3                  $4.23               $12.70
Ialallo29                   3                  $3.95               $11.84
Tyidaim51                   3                  $3.94               $11.83
Lassilsala30                3                  $3.84               $11.51
Chadolyla44                 3                  $3.82               $11.46
Iri67                       3                  $3.79               $11.37
Inguron55                   3                  $3.70               $11.11
Siallylis44                 3                  $3.46               $10.37
Saistyphos30                3                  $3.44               $10.33
Phyali88                    3                  $3.43               $10.30
Strithenu87                 3                  $3.39               $10.18
Lisim78                     3                  $3.34               $10.02
Haillyrgue51                3                  $3.17                $9.50
Phistym51                   2                  $4.75                $9.50
Lamil79                     2                  $4.64                $9.29
Aina42                      3                  $3.07                $9.22
Saesrideu94                 2                  $4.59                $9.18
Arin32                      2                  $4.54                $9.09
Rarallo90                   3                  $3.02                $9.05
Baelollodeu94               2                  $4.51                $9.03
Aelin32                     3                  $2.99                $8.98
Lisopela58                  3                  $2.95                $8.86
Saedaiphos46                3                  $2.94                $8.83
Reunasu60                   2                  $4.41                $8.82
Chanastnya43                3                  $2.94                $8.82
...                       ...                    ...                  ...
Lirtim36                    1                  $1.33                $1.33
Shidai42                    1                  $1.33                $1.33
Philodil43                  1                  $1.29                $1.29
Eyista89                    1                  $1.29                $1.29
Mindilsa60                  1                  $1.28                $1.28
Tyarithn67                  1                  $1.14                $1.14
Aisur51                     1                  $1.14                $1.14
Ilista82                    1                  $1.14                $1.14
Jiskimsda56                 1                  $1.14                $1.14
Aillyriadru65               1                  $1.10                $1.10
Eryon48                     1                  $1.10                $1.10
Iduelis31                   1                  $1.10                $1.10
Yasur85                     1                  $1.10                $1.10
Undjask33                   1                  $1.10                $1.10
Lirtassa52                  1                  $1.09                $1.09
Aelidru27                   1                  $1.09                $1.09
Yalaeria91                  1                  $1.06                $1.06
Aesurstilis64               1                  $1.03                $1.03
Saida58                     1                  $1.03                $1.03
Euthe35                     1                  $1.03                $1.03
Hala31                      1                  $1.02                $1.02
Qilalista41        

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [46]:
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Range  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [47]:
# This returns an object with multiple indexes, however, which can be harder to deal with
item_groupby_df = purchase_data_df.groupby(['Item ID', 'Item Name'])

item_groupby_df

In [48]:
item_purchase_count = item_groupby_df["Purchase ID"].count()

item_purchase_count.head()

Item ID  Item Name         
0        Splinter              4
1        Crucifer              3
2        Verdict               6
3        Phantomlight          6
4        Bloodlord's Fetish    5
Name: Purchase ID, dtype: int64

In [49]:
item_purchase_avg = item_groupby_df["Price"].mean()

sn_purchase_avg.head()

SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [50]:
item_purchase_total = item_groupby_df["Price"].sum()

item_purchase_total.head()

Item ID  Item Name         
0        Splinter               5.12
1        Crucifer               9.78
2        Verdict               14.88
3        Phantomlight          14.94
4        Bloodlord's Fetish     8.50
Name: Price, dtype: float64

In [51]:
popular_items_df = pd.DataFrame({"Purchase Count" : item_purchase_count, "Item Price" : item_purchase_avg,
                                "Total Purchase Value": item_purchase_total})
popular_items_df.sort_index().round(2)

Purchase Count  \
Item ID Item Name                                                      
0       Splinter                                                   4   
1       Crucifer                                                   3   
2       Verdict                                                    6   
3       Phantomlight                                               6   
4       Bloodlord's Fetish                                         5   
5       Putrid Fan                                                 4   
6       Rusty Skull                                                2   
7       Thorn, Satchel of Dark Souls                               7   
8       Purgatory, Gem of Regret                                   3   
9       Thorn, Conqueror of the Corrupted                          4   
10      Sleepwalker                                                4   
11      Brimstone                                                  5   
12      Dawne                                                      6   
13      Serenity                                                   4   
14      Possessed Core                                             2   
15      Soul Infused Crystal                                       5   
16      Restored Bauble                                            4   
17      Lazarus, Terror of the Earth                               4   
18      Torchlight, Bond of Storms                                 2   
19      Pursuit, Cudgel of Necromancy                              8   
20      Netherbane                                                 5   
21      Souleater                                                  5   
22      Amnesia                                                    6   
23      Crucifer                                                   1   
24      Warped Fetish                                              5   
25      Hero Cane                                                  6   
26      Unholy Wand                                                2   
27      Riddle, Tribute of Ended Dreams                            1   
28      Flux, Destroyer of Due Diligence                           2   
29      Chaos, Ender of the End                                    5   
...                                                              ...   
154     Feral Katana                                               4   
155     War-Forged Gold Deflector                                  5   
156     Soul-Forged Steel Shortsword                               5   
157     Spada, Etcher of Hatred                                    4   
158     Darkheart, Butcher of the Champion                         2   
159     Oathbreaker, Spellblade of Trials                          7   
160     Azurewrath                                                 6   
161     Devine                                                     5   
162     Abyssal Shard                                              5   
163     Thunderfury Scimitar                                       5   
164     Exiled Doomblade                                           7   
165     Bone Crushing Silver Skewer                                4   
166     Thirsty Iron Reaver                                        3   
167     Malice, Legacy of the Queen                                4   
168     Sun Strike, Jaws of Twisted Visions                        2   
169     Interrogator, Blood Blade of the Queen                     4   
170     Shadowsteel                                                4   
171     Scalpel                                                    3   
172     Blade of the Grave                                         4   
173     Stormfury Longsword                                        2   
174     Primitive Blade                                            4   
175     Woeful Adamantite Claymore                                 3   
176     Relentless Iron Skewer                                     2   
177     Winterthorn, D

In [52]:
#Sort by the purchase count column in descending order
popular_items_df = popular_items_df.sort_values(by='Purchase Count', ascending=False)
popular_items_df

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   
103     Singed Scalpel                                             8   
75      Brutality Ivory Warmace                                    8   
72      Winter's Bite                                              8   
60      Wolf                                                       8   
59      Lightning, Etcher of the King                              8   
37      Shadow Strike, Glory of Ending Hope                        8   
34      Retribution Axe                                            8   
92      Final Critic                                               8   
53      Vengeance Cleaver                                          7   
110     Suspension                                                 7   
7       Thorn, Satchel of Dark Souls                               7   
71      Demise                                                     7   
117     Heartstriker, Legacy of the Light                          7   
159     Oathbreaker, Spellblade of Trials                          7   
85      Malificent Bag                                             7   
164     Exiled Doomblade                                           7   
141     Persuasion                                                 7   
78      Glimmer, Ender of the Moon                                 7   
102     Avenger                                                    6   
129     Fate, Vengeance of Eternal Justice                         6   
93      Apocalyptic Battlescythe                                   6   
54      Eternal Cleaver                                            6   
3       Phantomlight                                               6   
2       Verdict                                                    6   
40      Second Chance                                              6   
...                                                              ...   
81      Dreamkiss                                                  2   
6       Rusty Skull                                                2   
18      Torchlight, Bond of Storms                                 2   
26      Unholy Wand                                                2   
28      Flux, Destroyer of Due Diligence                           2   
30      Stormcaller                                                2   
158     Darkheart, Butcher of the Champion                         2   
69      Frenzy, Defender of the Harvest                            2   
132     Persuasion                                                 2   
63      Stormfury Mace                                             2   
56      Foul Titanium Battle Axe                                   2   
127     Heartseeker, Reaver of Souls                               2   
48      Rage, Legacy of the Lone Victor                            2   
125     Whistling Mithril Warblade                                 2   
43      Foul Edge                                                  2   
115     Spectral Diamond Doomblade                                 2   
33      Curved Axe                                                 2   
31      Trickster                                                  2   
118     Ghost Reaver, Longsword of Magic                           1   
47      Alpha, Reach of Ending Hope                                1   
42      The Decapitator                                            1   
27      Riddle, Tribute of Ended Dreams                            1   
126     Exiled Mithril Longsword                                   1   
51      Endbringer    

In [53]:
#Sort the above table by the total purchase value column in descending order
popular_items_df = popular_items_df.sort_values(by='Total Purchase Value', ascending=False)
popular_items_df

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   
59      Lightning, Etcher of the King                              8   
108     Extraction, Quickblade Of Trembling Hands                  9   
78      Glimmer, Ender of the Moon                                 7   
72      Winter's Bite                                              8   
60      Wolf                                                       8   
179     Wolf, Promise of the Moonwalker                            6   
160     Azurewrath                                                 6   
25      Hero Cane                                                  6   
37      Shadow Strike, Glory of Ending Hope                        8   
139     Mercy, Katana of Dismay                                    5   
128     Blazeguard, Reach of Eternity                              5   
138     Peacekeeper, Wit of Dark Magic                             5   
141     Persuasion                                                 7   
87      Deluge, Edge of the West                                   5   
159     Oathbreaker, Spellblade of Trials                          7   
136     Ghastly Adamantite Protector                               6   
65      Conqueror Adamantite Mace                                  5   
101     Final Critic                                               5   
102     Avenger                                                    6   
133     Faith's Scimitar                                           5   
39      Betrayal, Whisper of Grieving Widows                       5   
111     Misery's End                                               4   
165     Bone Crushing Silver Skewer                                4   
52      Hatred                                                     4   
75      Brutality Ivory Warmace                                    8   
...                                                              ...   
124     Venom Claymore                                             4   
0       Splinter                                                   4   
155     War-Forged Gold Deflector                                  5   
48      Rage, Legacy of the Lone Victor                            2   
158     Darkheart, Butcher of the Champion                         2   
51      Endbringer                                                 1   
134     Undead Crusader                                            1   
30      Stormcaller                                                2   
135     Warped Diamond Crusader                                    3   
91      Celeste                                                    1   
177     Winterthorn, Defender of Shifting Worlds                   2   
70      Hope's End                                                 4   
115     Spectral Diamond Doomblade                                 2   
69      Frenzy, Defender of the Harvest                            2   
47      Alpha, Reach of Ending Hope                                1   
180     Stormcaller                                                1   
27      Riddle, Tribute of Ended Dreams                            1   
183     Dragon's Greatsword                                        3   
31      Trickster                                                  2   
168     Sun Strike, Jaws of Twisted Visions                        2   
90      Betrayer                                                   1   
33      Curved Axe                                                 2   
26      Unholy Wand                                                2   
118     Ghost Reaver, 

In [54]:
# Change the formatting of the columns to two decimals and add a $
popular_items_df["Item Price"]=popular_items_df["Item Price"].map('${:,.2f}'.format)
popular_items_df["Total Purchase Value"]=popular_items_df["Total Purchase Value"].map('${:,.2f}'.format)

In [55]:
#Display Final Formatted DataFrame
popular_items_df

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   
59      Lightning, Etcher of the King                              8   
108     Extraction, Quickblade Of Trembling Hands                  9   
78      Glimmer, Ender of the Moon                                 7   
72      Winter's Bite                                              8   
60      Wolf                                                       8   
179     Wolf, Promise of the Moonwalker                            6   
160     Azurewrath                                                 6   
25      Hero Cane                                                  6   
37      Shadow Strike, Glory of Ending Hope                        8   
139     Mercy, Katana of Dismay                                    5   
128     Blazeguard, Reach of Eternity                              5   
138     Peacekeeper, Wit of Dark Magic                             5   
141     Persuasion                                                 7   
87      Deluge, Edge of the West                                   5   
159     Oathbreaker, Spellblade of Trials                          7   
136     Ghastly Adamantite Protector                               6   
65      Conqueror Adamantite Mace                                  5   
101     Final Critic                                               5   
102     Avenger                                                    6   
133     Faith's Scimitar                                           5   
39      Betrayal, Whisper of Grieving Widows                       5   
111     Misery's End                                               4   
165     Bone Crushing Silver Skewer                                4   
52      Hatred                                                     4   
75      Brutality Ivory Warmace                                    8   
...                                                              ...   
124     Venom Claymore                                             4   
0       Splinter                                                   4   
155     War-Forged Gold Deflector                                  5   
48      Rage, Legacy of the Lone Victor                            2   
158     Darkheart, Butcher of the Champion                         2   
51      Endbringer                                                 1   
134     Undead Crusader                                            1   
30      Stormcaller                                                2   
135     Warped Diamond Crusader                                    3   
91      Celeste                                                    1   
177     Winterthorn, Defender of Shifting Worlds                   2   
70      Hope's End                                                 4   
115     Spectral Diamond Doomblade                                 2   
69      Frenzy, Defender of the Harvest                            2   
47      Alpha, Reach of Ending Hope                                1   
180     Stormcaller                                                1   
27      Riddle, Tribute of Ended Dreams                            1   
183     Dragon's Greatsword                                        3   
31      Trickster                                                  2   
168     Sun Strike, Jaws of Twisted Visions                        2   
90      Betrayer                                                   1   
33      Curved Axe                                                 2   
26      Unholy Wand                                                2   
118     Ghost Reaver, 

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [56]:
popular_items_df

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   
59      Lightning, Etcher of the King                              8   
108     Extraction, Quickblade Of Trembling Hands                  9   
78      Glimmer, Ender of the Moon                                 7   
72      Winter's Bite                                              8   
60      Wolf                                                       8   
179     Wolf, Promise of the Moonwalker                            6   
160     Azurewrath                                                 6   
25      Hero Cane                                                  6   
37      Shadow Strike, Glory of Ending Hope                        8   
139     Mercy, Katana of Dismay                                    5   
128     Blazeguard, Reach of Eternity                              5   
138     Peacekeeper, Wit of Dark Magic                             5   
141     Persuasion                                                 7   
87      Deluge, Edge of the West                                   5   
159     Oathbreaker, Spellblade of Trials                          7   
136     Ghastly Adamantite Protector                               6   
65      Conqueror Adamantite Mace                                  5   
101     Final Critic                                               5   
102     Avenger                                                    6   
133     Faith's Scimitar                                           5   
39      Betrayal, Whisper of Grieving Widows                       5   
111     Misery's End                                               4   
165     Bone Crushing Silver Skewer                                4   
52      Hatred                                                     4   
75      Brutality Ivory Warmace                                    8   
...                                                              ...   
124     Venom Claymore                                             4   
0       Splinter                                                   4   
155     War-Forged Gold Deflector                                  5   
48      Rage, Legacy of the Lone Victor                            2   
158     Darkheart, Butcher of the Champion                         2   
51      Endbringer                                                 1   
134     Undead Crusader                                            1   
30      Stormcaller                                                2   
135     Warped Diamond Crusader                                    3   
91      Celeste                                                    1   
177     Winterthorn, Defender of Shifting Worlds                   2   
70      Hope's End                                                 4   
115     Spectral Diamond Doomblade                                 2   
69      Frenzy, Defender of the Harvest                            2   
47      Alpha, Reach of Ending Hope                                1   
180     Stormcaller                                                1   
27      Riddle, Tribute of Ended Dreams                            1   
183     Dragon's Greatsword                                        3   
31      Trickster                                                  2   
168     Sun Strike, Jaws of Twisted Visions                        2   
90      Betrayer                                                   1   
33      Curved Axe                                                 2   
26      Unholy Wand                                                2   
118     Ghost Reaver, 

Description of three observable trends based on the data:

1. The Item "Oathbreaker, Last Hope of the Breaking Storm" was the single hishgest Total Purchasing Value. We can expect this item to be a high performing item in the future.

2. No gamer made more than 5 total purchases or spent more than $18.96 total on items. $20 per person in game revenue from items should be our maximum goal per person. 

3. Players purchased from players in the 20-24 age range made up over 45% of the total revenue earned from the sale of items. Therefore, this age range should be our target demographic.